torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1229,20 +1250,20 @@ def _optim_state_dict(
         group,
         param_id_to_param,
         param_to_fqns,
+        fqn_to_fsdp_param_info,
     )
-    flat_param_to_fsdp_state = _get_flat_param_to_fsdp_module(model)
 
     # Iterate in rank 0's flattened parameter ID order to ensure aligned
     # all-gathers across ranks
     for optim_state_key in param_id_to_optim_state_key.values():
         param_id = optim_state_key_to_param_id[optim_state_key]
-        if optim_state_key.is_flat_param:
-            param = param_id_to_param[param_id]
-            fsdp_state = flat_param_to_fsdp_state[param]
+        if optim_state_key.is_fsdp_wrapped:
+            fqn = optim_state_key.unflat_param_names[0]
```

Comment:
Conceptual: Why do we only take 0th index here?

IIUC, `unflat_param_names` is a `list` of the FQNs corresponding to 1 `FlatParameter`. Do we take the 0th index because `fqn_to_fsdp_param_info` is a surjective map, where we can key by any FQN corresponding to the `FlatParameter` to get the same `FSDPParamInfo`?

Reply:
If there are multiple `unflat_param_names` (not use_orig_params), they share the same FSDPParamInfo. So the first `unflat_param_name` is sufficient to fetch the FSDPParamInfo for both use or not use orig_params cases.